src/lua/lua_tcp.c

```diff
@@ -1494,6 +1969,226 @@ lua_tcp_shift_callback (lua_State *L)
+	}
+
+	struct thread_entry *thread = lua_thread_pool_get_running_entry (cbd->cfg->lua_thread_pool);
+
+	rh = g_malloc0 (sizeof (*rh));
+	rh->type = LUA_WANT_READ;
+	rh->h.r.cbref = -1;
+
+	msg_debug_tcp ("added read sync event, thread: %p", thread);
+
+	g_queue_push_tail (cbd->handlers, rh);
+	lua_tcp_plan_handler_event (cbd, TRUE, TRUE);
+
+	TCP_RETAIN (cbd);
+
+	return lua_thread_yield (thread, 0);
+}
+
+static int
+lua_tcp_sync_write (lua_State *L)
+{
+	LUA_TRACE_POINT;
+	struct lua_tcp_cbdata *cbd = lua_check_sync_tcp (L, 1);
+	struct lua_tcp_handler *wh;
+	gint tp;
+	struct iovec *iov = NULL;
+	guint niov = 0;
+	gsize total_out = 0;
+
+	if (cbd == NULL) {
+		return luaL_error (L, "invalid arguments [self is not rspamd{tcp_sync}]");
+	}
+
+	struct thread_entry *thread = lua_thread_pool_get_running_entry (cbd->cfg->lua_thread_pool);
+
+	tp = lua_type (L, 2);
+	if (tp == LUA_TSTRING || tp == LUA_TUSERDATA) {
+		iov = g_malloc (sizeof (*iov));
+		niov = 1;
+
+		if (!lua_tcp_arg_toiovec (L, 2, cbd, iov)) {
+			msg_err ("tcp request has bad data argument");
+			g_free (iov);
+			g_free (cbd);
+
+			return luaL_error (L, "invalid arguments second parameter (data) is expected to be either string or rspamd{text}");
+		}
+
+		total_out = iov[0].iov_len;
+	}
+	else if (tp == LUA_TTABLE) {
+		/* Count parts */
+		lua_pushvalue (L, 3);
+
+		lua_pushnil (L);
+		while (lua_next (L, -2) != 0) {
+			niov ++;
+			lua_pop (L, 1);
+		}
+
+		iov = g_malloc (sizeof (*iov) * niov);
+		lua_pushnil (L);
+		niov = 0;
+
+		while (lua_next (L, -2) != 0) {
+			if (!lua_tcp_arg_toiovec (L, -1, cbd, &iov[niov])) {
+				msg_err ("tcp request has bad data argument at pos %d", niov);
+				g_free (iov);
+				g_free (cbd);
+
+				return luaL_error (L, "invalid arguments second parameter (data) is expected to be either string or rspamd{text}");
+			}
+
+			total_out += iov[niov].iov_len;
+			niov ++;
+
+			lua_pop (L, 1);
+		}
+
+		lua_pop (L, 1);
+	}
+
+	wh = g_malloc0 (sizeof (*wh));
+	wh->type = LUA_WANT_WRITE;
+	wh->h.w.iov = iov;
+	wh->h.w.iovlen = niov;
+	wh->h.w.total_bytes = total_out;
+	wh->h.w.pos = 0;
+	wh->h.w.cbref = -1;
+	msg_debug_tcp ("added sync write event, thread: %p", thread);
+
+	g_queue_push_tail (cbd->handlers, wh);
+	lua_tcp_plan_handler_event (cbd, TRUE, TRUE);
+
+	TCP_RETAIN (cbd);
+	// lua_tcp_register_event (cbd);
+
+	return lua_thread_yield (thread, 0);
+}
+
+static gint
+lua_tcp_sync_eof(lua_State *L)
+{
+	LUA_TRACE_POINT;
+	struct lua_tcp_cbdata *cbd = lua_check_sync_tcp (L, 1);
+	if (cbd == NULL) {
+		return luaL_error (L, "invalid arguments [self is not rspamd{tcp_sync}]");
+	}
+
+	lua_pushboolean(L, cbd->eof);
+
+	return 1;
+}
+
+static gint
+lua_tcp_sync_shutdown (lua_State *L)
+{
+	LUA_TRACE_POINT;
+	struct lua_tcp_cbdata *cbd = lua_check_sync_tcp (L, 1);
+	if (cbd == NULL) {
+		return luaL_error (L, "invalid arguments [self is not rspamd{tcp_sync}]");
+	}
+
+	shutdown (cbd->fd, SHUT_WR);
```

Comment:
It differs from the standard unix shutdown semantics. I don't know, maybe leave it as is for now but consider adding `SHUT_RD` as an optional argument.